File: src/main/kotlin/wetfe/core/character/CharacterManager.kt
package wetfe.core.character

//
//class CharacterDocumentManager {
//    private _documents: CharacterDocument[] = [];
//    private _map: Map<string, CharacterDocument> = new Map();
//    private _active_cdoc: CharacterDocument;
//    private _target_cdoc: CharacterDocument;
//
//    constructor() {
//        this._active_cdoc = new CharacterDocument("a_id", "a_folder_id", new Character(new CharData()));
//        this._target_cdoc = new CharacterDocument("t_id", "t_folder_id", new Character(new CharData()));
//    }
//
//    get docs() {
//        return this._documents;
//    }
//
//    get active() {
//        return this._active_cdoc.character;
//    }
//
//    get target() {
//        return this._target_cdoc.character;
//    }
//
//    get a_cdoc() {
//        return this._active_cdoc;
//    }
//
//    get t_cdoc() {
//        return this._target_cdoc;
//    }
//
//    get a_char() {
//        return this.active;
//    }
//
//    get t_char() {
//        return this.target;
//    }
//
//    get a_data() {
//        return this._active_cdoc.data;
//    }
//
//    get t_data() {
//        return this._target_cdoc.data;
//    }
//
//    get a_state() {
//        return this._active_cdoc.state;
//    }
//
//    get t_state() {
//        return this._target_cdoc.state;
//    }
//
//    getCharacterDocuments() {
//        return this.docs;
//    }
//
//    getActiveCharacterDocument() {
//        return this.a_cdoc;
//    }
//
//    getTargetCharacterDocument() {
//        return this.t_cdoc;
//    }
//
//    getActiveCharacter() {
//        return this.active;
//    }
//
//    getTargetCharacter() {
//        return this.target;
//    }
//
//    getActiveData() {
//        return this.a_data;
//    }
//
//    getTargetData() {
//        return this.t_data;
//    }
//
//    getActiveState() {
//        return this.a_state;
//    }
//
//    getTargetState() {
//        return this.t_state;
//    }
//
//    setActive(c: CharacterDocument | Character | CharData | string) {
//        const ckey = typeof c === 'string' ? c : c.key;
//        const cdoc = this._map.get(ckey);
//        if (cdoc) {
//            this._active_cdoc = cdoc;
//        }
//        return this._active_cdoc.character;
//    }
//
//    setTarget(c: CharacterDocument | Character | CharData | string) {
//        const ckey = typeof c === 'string' ? c : c.key;
//        const cdoc = this._map.get(ckey);
//        if (cdoc) {
//            this._target_cdoc = cdoc;
//        }
//        return this._target_cdoc.character;
//    }
//
//    add(cdoc: CharacterDocument) {
//        if (!this._map.has(cdoc.key)) {
//            this._documents.push(cdoc);
//            this._map.set(cdoc.key, cdoc);
//        }
//    }
//}
//const CDM = new CharacterDocumentManager();
//
//function setEditMode(on: boolean) {
//    if (on) {
//        $('.we-edit-mode').removeClass('d-none');
//        $('#edit-mode-btn').addClass('d-none');
//    } else {
//        $('.we-edit-mode').addClass('d-none');
//        $('#edit-mode-btn').removeClass('d-none');
//    }
//}
//
//function toggleEditModeNavSpan() {
//    const $emns = $('#edit-mode-nav-span');
//    if ($emns.is('.d-none')) {
//        $emns.removeClass('d-none');
//    } else {
//        $emns.addClass('d-none');
//    }
//}
//
//function toggleEditMode() {
//    const $em_elements = $('.we-edit-mode');
//    if ($em_elements.is('.d-none')) {
//        $em_elements.removeClass('d-none');
//    } else {
//        $em_elements.addClass('d-none');
//    }
//}
//
//function increment(k: string, n: number, is_state = STATE_PARAM_MAP.has(k)) {
//    let val: number;
//    let stateparam: StateParam;
//    let charparam: CoreParam;
//    if (is_state) {
//        const sparam = STATE_PARAM_MAP.get(k);
//        if (sparam === undefined)
//            throw new Error('Unable to lookup StateParam for type ' + k);
//        val = CDM.active.getStateVal(sparam);
//        stateparam = sparam;
//        charparam = DEFAULT_CORE_PARAM;
//    } else {
//        const cparam = CORE_PARAM_MAP.get(k);
//        if (cparam === undefined)
//            throw new Error('Unable to lookup CharParam for type ' + k);
//        val = CDM.active.getCoreVal(cparam);
//        charparam = cparam;
//        stateparam = DEFAULT_STATE_PARAM;
//    }
//    const data_prefix = is_state ? 'data-state-' : 'data-param-';
//    if (typeof val === 'number') {
//        let max = ATTR.max(k);
//        let wal = ATTR.bound(k, val + n);
//        if (wal === val) return;
//        if (is_state) {
//            CDM.active.setStateVal(stateparam, wal);
//        } else {
//            CDM.active.addCoreMod(charparam, new Modification((wal - val), 'FREE EDIT'));
//            if (wal >= 0 && wal < ATTR.NATURES.length)
//            $('.we-nature[data-nature-of="' + k + '"]').html(ATTR.natureOf(wal));
//        }
//        $('[' + data_prefix + 'key="' + k + '"]').html(String(wal));
//        let $linked_pbars = $('.progress-bar[' + data_prefix + 'link="' + k + '"]');
//        if ($linked_pbars.length > 0 && max > 0)
//        $linked_pbars.css('width', '' + (100 * wal / max).toFixed(0) + '%');
//    }
//}
//
//function incrementElement(element: Element, n = 1) {
//    let $element = $(element);
//    let is_state = $element.is('.we-state-change');
//    let key = is_state ?
//    $element.attr('data-state-target') :
//    $element.attr('data-param-target');
//    if (key === undefined)
//        throw new Error('Unable to extract a state or param target key from the given element.');
//    return increment(key, n, is_state);
//}
//
//function decrementElement(element: Element, n = 1) {
//    incrementElement(element, -n);
//}
//
//function decrement(k: string, n: number, is_state: boolean) {
//    increment(k, -n, is_state);
//}
//
//function reallocate(src_key?: string, dst_key?: string, n = 1) {
//    if (!src_key && !dst_key) return;
//    if (n < 1) return;
//    if (src_key) {
//        const srcparam = STATE_PARAM_MAP.get(src_key);
//        if (srcparam === undefined)
//            throw new Error('Unable to lookup state parameter for key ' + src_key);
//        let src_val = CDM.active.getStateVal(srcparam);
//        if (src_val < 1) return;
//        if (n > src_val) n = src_val;
//        decrement(src_key, n, true);
//    }
//    if (dst_key) {
//        increment(dst_key, n, true);
//    }
//}
//
//function takeDamage(n: number) {
//    reallocate('HEL', 'DMG', n);
//}
//
//function healDamage(n: number) {
//    reallocate('DMG', 'HEL', n);
//}
//
//function accumulateFatigue(src_key: string, n: number) {
//    if (typeof src_key !== 'string') src_key = 'HEL';
//    reallocate(src_key, 'FTG', n);
//}
//
//function restoreFatigue(n: number) {
//    reallocate('FTG', 'HEL', n);
//}
//
//function showCharPicker() {
//    $('.we-char-picker').removeClass('d-none');
//}
//
//function hideCharPicker() {
//    $('.we-char-picker').addClass('d-none');
//}
//
//function drawLoadingHud(msg: string) {
//    hideCharPicker();
//    $('.we-loading-spinner').addClass('fa-spin');
//    $('.we-loading-txt').html(msg);
//    $('.we-loading-hud').removeClass('d-none');
//}
//
//function clearLoadingHud() {
//    $('.we-loading-spinner').removeClass('fa-spin');
//    $('.we-loading-hud').addClass('d-none');
//    showCharPicker();
//}
//
//let $char_dditem_ce: JQuery | null = null;
//function addCharDropdownItem(char: Character) {
//    if ($char_dditem_ce === null) {
//        $char_dditem_ce = $('.we-char-dditem.we-clone-element');
//    }
//    $char_dditem_ce.clone()
//    .attr('data-char-key', char.key)
//            .removeClass('we-clone-element d-none')
//            .html(char.data.common_name)
//            .appendTo('.we-char-ddmenu');
//}
//
//function repopulateCharDropdown() {
//    $('.we-char-ddmenu>.we-char-dditem').remove();
//    for (let cdoc of CDM.docs) {
//        addCharDropdownItem(cdoc.character);
//    }
//}
//
//function handleCharacterCreated(cdoc: CharacterDocument) {
//    console.log('[CharManager::onCharacterCreated] INFO: Created character "' + cdoc.key + '"');
//    addCharacter(cdoc);
//    clearLoadingHud();
//    $('.we-char-dditem[data-char-key="' + cdoc.key + '"]').click();
//}
//
//function createCharacter(element: HTMLElement) {
//    //@ts-ignore
//    google.script.run.withSuccessHandler(handleCharacterCreated).createCharacter();
//    drawLoadingHud('Creating New Character');
//}
//
//function handleDataSaved() {
//    setEditMode(false);
//    clearLoadingHud();
//    $('#navbar-toggler').click();
//}
//
//function saveCurrentData() {
//    //@ts-ignore
//    google.script.run.withSuccessHandler(handleDataSaved).saveCharacter(CDM.a_cdoc);
//    drawLoadingHud('Saving Character');
//}
//
//function addCharacter(cdoc: CharacterDocument) {
//    if (!cdoc) {
//        console.error('[CharManager::addCharacter] ERROR: The given character "char" was null.');
//        return;
//    }
//    CDM.add(cdoc);
//    addCharDropdownItem(cdoc.character);
//}
//
//function drawCharacter() {
//    const chara = CDM.active;
//    const cstate = CDM.a_state;
//    $('.we-character-name').html(chara.data.full_name);
//    for (let [ck, cv] of CORE_PARAM_MAP) {
//        let val = chara.getCoreVal(cv);
//        $('[data-param-key="' + ck + '"]').html(val.toString());
//        if (typeof val === 'number') {
//        let $linked_pbars = $('.progress-bar[data-param-link="' + ck + '"]');
//        if ($linked_pbars.length > 0) {
//        let max = ATTR.max(ck);
//        $linked_pbars.css('width', '' + (100 * val / max).toFixed(0) + '%');
//    }
//        let $natures = $('.we-nature[data-nature-of="' + ck + '"]');
//        if ($natures.length > 0) {
//        $natures.html(ATTR.natureOf(val));
//    }
//    }
//    }
//    for (let [sk, sv] of STATE_PARAM_MAP) {
//        let val = cstate.of(sv);
//        $('[data-state-key="' + sk + '"]').html(val.toString());
//        let $linked_pbars = $('.progress-bar[data-state-link="' + sk + '"]');
//        if ($linked_pbars.length > 0) {
//        let max = ATTR.max(sk);
//        $linked_pbars.css('width', '' + (100 * val / max).toFixed(0) + '%');
//    }
//    }
//    // TODO: display the active character's condition (cstate.condition)
//}
//
//function selectCharacter(element: HTMLElement) {
//    let $cdditem = $(element);
//    let ckey = $cdditem.attr('data-char-key');
//    if (typeof ckey !== 'string') {
//        console.log('[CharManager::selectCharacter] ERROR: char dditem missing attribute "data-char-key"');
//        return;
//    }
//    CDM.setActive(ckey);
//    drawCharacter();
//}
//
//function handleLoadedCharacterDocuments(cdocs: CharacterDocument[]) {
//    if (Array.isArray(cdocs)) {
//        for (let c of cdocs) {
//            addCharacter(c);
//        }
//    }
//    clearLoadingHud();
//    //$('#data-serialization-text').val(JSON.stringify(CHARS));
//}
//
//function loadCharacterDocuments() {
//    //@ts-ignore
//    google.script.run.withSuccessHandler(handleLoadedCharacterDocuments).getCharacterDocuments();
//    drawLoadingHud(`Loading Character Documents`);
//}
//
//$(function () {
//    $('[data-toggle="tooltip"]').tooltip();
//    loadCharacterDocuments();
//});

File: src/main/resources/static/character-manager.js
/*
* !THIS FILE MUST BE LOADED AFTER kotlin.js AND wetfe.js
*/
if (typeof wetfe === 'undefined') {
    throw new Error("Error loading script 'character-manager.js'. Its dependency 'wetfe' was not found. Please, check whether 'wetfe' is loaded prior to 'encounter.js'.");
}

let charaMap = {};
let character = null; // the current character

const CORE = wetfe.wetfe.core;
const QuadStat = CORE.character.QuadStat;
const CoreParam = CORE.character.CoreParam;
const StateParam = CORE.character.StateParam;
const StatMode = CORE.character.StatMode;
const Attribute = CORE.character.Attribute;
const CharaData = CORE.character.CharaData;
const Character = CORE.character.Character;

function rand(min, max) {
    return Math.random() * (max - min) + min;
}

function clone(obj) {
    return JSON.parse(JSON.stringify(obj));
}

function setEditMode(on) {
    if (on) {
        $('.we-edit-mode').removeClass('d-none');
        $('#edit-mode-btn').addClass('d-none');
    } else {
        $('.we-edit-mode').addClass('d-none');
        $('#edit-mode-btn').removeClass('d-none');
    }
    const $form = $('#we-chara-edit-form');
    $form.empty();
    Object.getOwnPropertyNames(character.cdata_wnkotf$_0).forEach(function(p) {
        const v = character.cdata_wnkotf$_0[p];
        if (v === null) return;
        if (typeof v === 'string') {
            const newId = 'we-chara-field-text-' + p;
            const $newField = $('.we-chara-edit-text.we-clone-element').clone();
            $newField.children('label').attr('for', newId).text(p);
            $newField.children('div').children('textarea').attr('id', newId).text(v);
            $newField.removeClass('we-clone-element d-none').appendTo($form);
        } else {
            console.log(`TODO: handle character field "${p}" of type "${typeof v}"`)
        }
    });
}

function toggleEditModeNavSpan() {
    var $emns = $('#edit-mode-nav-span');
    if ($emns.is('.d-none')) $emns.removeClass('d-none');
    else $emns.addClass('d-none');
}

function toggleEditMode() {
    var $em_elements = $('.we-edit-mode');
    if ($em_elements.is('.d-none')) $em_elements.removeClass('d-none');
    else $em_elements.addClass('d-none');
}

function INC(k, n, is_state) {
    if (typeof is_state === 'undefined' || is_state === null) is_state = isSTATE(k);
    var data_prefix = is_state  ? 'data-state-' : 'data-param-';
    var val = is_state ? STATE[k](character) : CHAR[k](character);
    if (typeof val === 'number') {
        var max = MAX(k);
        var wal = BOUND(k, val + n);
        if (wal === val) return;
        if (is_state) {
            STATE[k](character, wal);
        } else {
            var mod = {
                type: "FREE EDIT",
                level: CHAR.LVL(character),
                val: (wal - val)
            };
            CHAR[k](character, mod);
            if (wal >= 0 && wal < NATURES.length) $('.we-nature[data-nature-of="' + k + '"]').html(NATURES[wal]);
        }
        $('[' + data_prefix + 'key="' + k + '"]').html(wal);
        var $linked_pbars = $('.progress-bar[' + data_prefix + 'link="' + k + '"]');
        if ($linked_pbars.length > 0 && isPOS(max)) $linked_pbars.css('width', '' + (100*wal/max).toFixed(0) + '%');
    }
}

function increment(element, n) {
    if (!character) return;
    if (typeof n !== 'number') n = 1;
    var $element = $(element);
    var is_state = $element.is('.we-state-change');
    var key = is_state ? $element.attr('data-state-target') : $element.attr('data-param-target');
    return INC(key, n, is_state);
}

function decrement(element, n) {
    if (typeof n !== 'number') n = 1;
    increment(element, -n);
}

function DEC(k, n, is_state) {
    INC(k, -n, is_state);
}

function reallocate(src_key, dst_key, n) {
    if (!(character && character.state)) return;
    if (!src_key && !dst_key) return;
    if (typeof n !== 'number') n = 1;
    if (n < 1) return;
    if (typeof src_key === 'string') {
        var src_val = STATE[src_key](character);
        if (src_val < 1) return;
        if (n > src_val) n = src_val;
        DEC(src_key, n, true);
    }
    if (typeof dst_key === 'string') {
        INC(dst_key, n, true);
    }
}

function takeDamage(n) {
    reallocate('HEL', 'DMG', n);
}

function healDamage(n) {
    reallocate('DMG', 'HEL', n);
}

function accumulateFatigue(src_key, n) {
    if (typeof src_key !== 'string') src_key = 'HEL';
    reallocate(src_key, 'FTG', n);
}

function restoreFatigue(n) {
    reallocate('FTG', 'HEL', n);
}

function showCharPicker() {
    $('.we-char-picker').removeClass('d-none');
}

function hideCharPicker() {
    $('.we-char-picker').addClass('d-none');
}

function drawLoadingHud(msg) {
    hideCharPicker();
    $('.we-loading-spinner').addClass('fa-spin');
    $('.we-loading-txt').html(msg);
    $('.we-loading-hud').removeClass('d-none');
}

function clearLoadingHud() {
    $('.we-loading-spinner').removeClass('fa-spin');
    $('.we-loading-hud').addClass('d-none');
    showCharPicker();
}

var $char_dditem_ce = null;
function addCharDropdownItem(char) {
    if ($char_dditem_ce === null) $char_dditem_ce = $('.we-char-dditem.we-clone-element');
    $char_dditem_ce.clone()
        .attr('data-chara-id', char.getId())
        .removeClass('we-clone-element d-none')
        .html(char.getName())
        .appendTo('.we-char-ddmenu');
}

function repopulateCharDropdown() {
    $('.we-char-ddmenu>.we-char-dditem').remove();
    if (Array.isArray(CHARS)) {
        for (var char of CHARS) {
            addCharDropdownItem(char);
        }
    }
}

// function onCharacterCreated(char) {
//     STATE.newState(char.data);
//     console.log('[PCMScript::onCharacterCreated] INFO: Created character "' + char.getId() + '"');
//     addCharacter(char);
//     clearLoadingHud();
//     $('.we-char-dditem[data-chara-id="' + char.getId() + '"]').click();
// }

// function createCharacter(element) {
//     drawLoadingHud('Creating New Character');
//     google.script.run.withSuccessHandler(onCharacterCreated).createCharacter();
// }

function createCharacter(element) {
        let newCharacter = new Character();
        console.log('[character-manager.js] INFO: Created character "' + newCharacter.getId() + '"');
        addCharacter(newCharacter);
        $('.we-character-name').text(newCharacter.getName());
        $('.we-char-dditem[data-chara-id="' + newCharacter.getId() + '"]').click();
}

function onDataSaved() {
    setEditMode(false);
    clearLoadingHud();
    $('#navbar-toggler').click();
}

function saveCurrentData() {
    if (!character) return;
    // drawLoadingHud('Saving Character');
    console.log(`TODO: Save the following string of character data:\n${JSON.stringify(character)}`)
}

function addCharacter(char) {
    if (!char) {
        console.error('[character-manager.js] ERROR: The given character was null.');
        return;
    }
    if (!charaMap || typeof charaMap !== 'object') charaMap = {};
    charaMap[char.getId()] = char;
    addCharDropdownItem(char);
}

function drawCharacter() {
    if (!character) return;
    $('.we-character-name').html(character.getFullName());
    for (let cparam of CoreParam.values()) {
        let val = character.getParam_qosqn7$(cparam);
        let pk = cparam.key;
        console.log(`Drawing character core param ${pk} with val of ${val}`);
        $('[data-param-key="' + pk + '"]').html(val);
        if (typeof val === 'number') {
            let $linked_pbars = $('.progress-bar[data-param-link="' + pk + '"]');
            if ($linked_pbars.length > 0) {
                let max = Attribute.Limit.Companion.of_qosqn7$(cparam).max;
                if (max > 0) $linked_pbars.css('width', '' + (100*val/max).toFixed(0) + '%');
            }
            let $natures = $('.we-nature[data-nature-of="' + pk + '"]');
            if ($natures.length > 0) {
                $natures.html(Attribute.Assessment.Companion.of_za3lpa$(val));
            }
        }
    }
    for (let sparam of StateParam.values()) {
        let val = character.getParam_alacmf$(sparam);
        let sk = sparam.key;
        console.log(`Drawing character state param ${sk} with val of ${val}`);
        $('[data-state-key="' + sk + '"]').html(val);
        let $linked_pbars = $('.progress-bar[data-state-link="' + sk + '"]');
        if ($linked_pbars.length > 0) {
            let max = Attribute.Limit.Companion.of_alacmf$(sparam);
            if (max > 0) $linked_pbars.css('width', '' + (100*val/max).toFixed(0) + '%');
        }
    }
}

function selectCharacter(element) {
    $cdditem = $(element);
    const cid = $cdditem.attr('data-chara-id');
    if (typeof cid !== 'string') {
        console.log('[character-manager.js] ERROR: char dditem missing attribute "data-chara-id"');
        return;
    }
    character = new Character(clone(charaMap[cid]).cdata_wnkotf$_0);
    if (!character) {
        console.log('[character-manager.js] ERROR: Unable to lookup character with id "' + cid + '"');
        return;
    }
    drawCharacter();
}

// function onDataLoaded(chars) {
//     CHARS = [];
//     if (Array.isArray(chars)) {
//         for (var c of chars) {
//             addCharacter(c);
//         }
//     }
//     clearLoadingHud();
//     //$('#data-serialization-text').val(JSON.stringify(CHARS));
// }

// function loadData() {
//     google.script.run.withSuccessHandler(onDataLoaded).getCharacters();
// }

$(function () {
    $('[data-toggle="tooltip"]').tooltip();
});


///

// CHARACTER OPERATORS
// const PARAM_KEYS = Object.freeze([
//     "LVL", "EXP", "SP", "CON", "DEX", "INT", "WIL"
// ]);
//
// function isPARAM(key) {
//     return PARAM_KEYS.includes(key);
// }

///



// CHARACTER STATE OPERATORS
// const STATE_KEYS = Object.freeze([
//     "CDN", "PWR", "HEL", "BRK", "DMG", "AFL", "FTG"
// ]);
//
// function isSTATE(key) {
//     return STATE_KEYS.includes(key);
// }
//
// function newState(cdata) {
//     cdata.state = {
//         condition: "NORMAL",
//         power: 0,
//         //health: CHAR.SP(cdata), //TODO: subtract soulbound chips
//         breakage: 0,
//         damage: 0,
//         affliction: 0,
//         fatigue: 0
//     };
// }
